flextape/service/prioritizer.go

```diff
@@ -0,0 +1,154 @@
+package service
+
+// Prioritizer is an object capable of sorting a queue in order of priority.
+//
+// In order for the prioritizer to have enough data to make the prioritization
+// decision, it must be invoked every time an entry is queued and dequeued, and
+// every time a license is allocated and released.
+type Prioritizer interface {
+	// Enqueued is called every time an invocation is queued.
+	Enqueued(inv *invocation)
+	// Dequeued is called every time an invocation is dequeued.
+	//
+	// Note that an invocation may be dequeued because it is expired,
+	// because the user withdrew the request, or because an allocation
+	// is now possible.
+	// In this last case, Dequeued() will be followed by an Allocated()
+	// call.
+	Dequeued(inv *invocation)
+
+	// Allocated is called every time an invocation is allocated a license.
+	Allocated(inv *invocation)
+	// Released is called every time an invocation loses a license.
+	//
+	// This generally means that the invocation was serviced and the
+	// license is no longer needed, but it could be called also if,
+	// for example, a client timed out while holding a license.
+	Released(inv *invocation)
+
+	// Sorter is a function that returns a function capable of
+	// reordering the queue.
+	//
+	// The returned Sorter is generally passed to invocationq.Sort.
+	Sorter() Sorter
+}
+
+// FIFOPrioritizer simply keeps the entries in the order they were queued.
+type FIFOPrioritizer struct {
+}
+
+func (_ *FIFOPrioritizer) Enqueued(inv *invocation) {
+}
+func (_ *FIFOPrioritizer) Dequeued(inv *invocation) {
+}
+func (_ *FIFOPrioritizer) Allocated(inv *invocation) {
+}
+func (_ *FIFOPrioritizer) Released(inv *invocation) {
+}
+func (_ *FIFOPrioritizer) Sorter() Sorter {
+	return nil
+}
+
+// EvenOwnerAllocationsPrioritizer is a prioritizer that tries to spread
+// license allocations evenly across owners.
+//
+// Time is not part of the equation: if user A has 10 licenses allocated
+// and an user B comes by, it will prioritize B over A until there is a
+// 5 ~ 5 distribution in number of licenses.
+type EvenOwnersPrioritizer struct {
+	// Key: invocation.Owner, value represents # of requests queued,
+	// monothonically increasing.
+	enqueued map[string]uint64
```

Comment:
Are we worried about race conditions/protecting these with mutexes? I suspect that a top-level mutex ends up protecting races here in practice, but on the flipside if mutexes were added here as appropriate probably the top-level mutex would result in the lower-level ones being uncontended and the performance impact being negligible.

Reply:
I'd rather not add locking here, reasons:
1. the `license` object has no locking, it is not thread safe by contract. It is made thread safe by the caller, service, which has a mutex protecting it. license uses different data structures that are not thread safe, this would be just another one of them, and having a top level caller ensure thread safety is a common pattern in go. Having just this one thread safe would make the code harder to read/update (when reading the code: "why is this data structure thread safe and the others are not? let me find where in the code it is run from multiple threads!" - a bit like using sync.Map or atomic in a class that otherwise does not look like it is used by multiple threads).
2. the `Sorter()` method has to access the content of the invocation. Is it safe to access? does it need a lock on invocation? or is each invocation already locked by the caller? Can this cause lock ordering deadlock? ... for the prioritizer to be thread safe / have useful locking, invocation has to have some thread safety guarantee - probably a lock per invocation? This change would be necessary, but imho out of scope for the PR. IF we say the caller guarantees invocation safety, then I think it would be best to have all methods expect a similar contract?